istioctl/cmd/tag.go

```diff
@@ -0,0 +1,387 @@
+// this retrieves the webhook created at revision installation rather than tag webhooks
+func getWebhooksWithRevision(ctx context.Context, client kubernetes.Interface, rev string) ([]admit_v1.MutatingWebhookConfiguration, error) {
+	webhooks, err := client.AdmissionregistrationV1().MutatingWebhookConfigurations().List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s,!%s", label.IstioRev, rev, istioTagLabel),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return webhooks.Items, nil
+}
+
+// getNamespacesWithTag retrieves all namespaces pointed at the given tag
+func getNamespacesWithTag(ctx context.Context, client kubernetes.Interface, tag string) ([]string, error) {
+	namespaces, err := client.CoreV1().Namespaces().List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", label.IstioRev, tag),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	nsNames := make([]string, len(namespaces.Items))
+	for i, ns := range namespaces.Items {
+		nsNames[i] = ns.Name
+	}
+	return nsNames, nil
+}
+
+// getTagName extracts tag name from webhook object
+func getTagName(wh admit_v1.MutatingWebhookConfiguration) (string, error) {
+	if tagName, ok := wh.ObjectMeta.Labels[istioTagLabel]; ok {
+		return tagName, nil
+	}
+	return "", fmt.Errorf("could not extract tag name from webhook")
+}
+
+// getRevision extracts tag target revision from webhook object
+func getTagRevision(wh admit_v1.MutatingWebhookConfiguration) (string, error) {
+	if tagName, ok := wh.ObjectMeta.Labels[label.IstioRev]; ok {
+		return tagName, nil
+	}
+	return "", fmt.Errorf("could not extract tag revision from webhook")
+}
+
+// deleteTagWebhooks deletes the given webhooks
+func deleteTagWebhooks(ctx context.Context, client kubernetes.Interface, webhooks []admit_v1.MutatingWebhookConfiguration) error {
+	var result error
+	for _, wh := range webhooks {
+		result = multierror.Append(client.AdmissionregistrationV1().MutatingWebhookConfigurations().Delete(ctx, wh.Name, metav1.DeleteOptions{})).ErrorOrNil()
+	}
+	return result
+}
+
+// buildDeleteTagConfirmation takes a list of webhooks and creates a message prompting confirmation for their deletion
+func buildDeleteTagConfirmation(tag string, taggedNamespaces []string) string {
+	var sb strings.Builder
+	base := fmt.Sprintf("Caution, found %d namespace(s) still pointing to tag \"%s\":", len(taggedNamespaces), tag)
+	sb.WriteString(base)
+	for _, ns := range taggedNamespaces {
+		sb.WriteString(" " + ns)
+	}
+	sb.WriteString("\nProceed with operation? [y/N]")
+
+	return sb.String()
+}
+
+// buildTagWebhookFromCanonical takes a canonical injector webhook for a given revision and generates a tag webhook
+// from the original webhook, we need to change (1) the namespace selector (2) the name (3) the labels
+func buildTagWebhookFromCanonical(wh admit_v1.MutatingWebhookConfiguration, tag, revision string) (*admit_v1.MutatingWebhookConfiguration, error) {
+	tagWebhook := new(admit_v1.MutatingWebhookConfiguration)
```

Comment:
I feel like the standard way to do this with structs is "struct{}" not "new(struct)", but I could be wrong.